tests/DataValues/TimeValueCalculatorTest.php

```diff
@@ -209,4 +209,279 @@ public function testGetSecondsForPrecision( $precision, $expected ) {
 		$this->assertEquals( $expected, $seconds );
 	}
 
+	/**
+	 * @return array
+	 */
+	private function timestampWithoutSignProvider() {
+		return [
+			'1-06-16T11:45:12Z',
+			'16-11-11T06:08:04Z',
+			'245-04-30T00:00:00Z',
+			'1054-02-11T14:00:02Z',
+			'2012-02-29T23:59:59Z',
+			#'2012-03-01T00:00:00Z',
+			'2013-12-31T23:59:59Z',
+			'2014-01-01T00:00:00Z',
+			'54844518-04-25T02:00:00Z',
+			'9990994999299999-07-09T14:20:00Z',
+		];
+	}
+
+	/**
+	 * @return array of precisions from the most to the least precise
+	 */
+	private function simplePrecisionProvider() {
+		return [
+			TimeValue::PRECISION_SECOND,
+			TimeValue::PRECISION_MINUTE,
+			TimeValue::PRECISION_HOUR,
+			TimeValue::PRECISION_DAY,
+			TimeValue::PRECISION_MONTH,
+			TimeValue::PRECISION_YEAR,
+			TimeValue::PRECISION_YEAR10,
+			TimeValue::PRECISION_YEAR100,
+			TimeValue::PRECISION_YEAR1K,
+			TimeValue::PRECISION_YEAR10K,
+			TimeValue::PRECISION_YEAR100K,
+			TimeValue::PRECISION_YEAR1M,
+			TimeValue::PRECISION_YEAR10M,
+			TimeValue::PRECISION_YEAR100M,
+			TimeValue::PRECISION_YEAR1G
+		];
+	}
+
+	/**
+	 * Check that:
+	 *   - timestampWithoutSignProvider() values are ordered from earliest to latest according
+	 *       to getLowerTimestamp(),
+	 *   - getTimestamp() values are always greater or equal than getLowerTimestamp() ones,
+	 *   - higher precisions of the same timestamp always correspond to greater or equal
+	 *       getLowerTimestamp() values than lower precisions,
+	 *   - higher before values for the same timestamp always correspond to lower or equal
+	 *       getLowerTimestamp() values than lower before values,
+	 *   - a few TimeValue values are equal to their expected getLowerTimestamp() ones.
+	 */
```

Comment:
I would prefer if these tests were split up into a lot more functions. For example:

* One test for checking specific `TimeValue`s and their expected timestamps, with a [`@dataProvider`](https://phpunit.de/manual/6.5/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers) (note that MediaWiki [prefers](https://www.mediawiki.org/wiki/Manual:PHP_unit_testing/Writing_unit_tests#Data_providers) names like `provideTimeValuesWithLowerTimestamps`) for the constants that are used below.
* One test for comparing the direct timestamp with the lower timestamp, with a data provider for timestamps.
* One test for comparing timestamps at different precisions, with the same data provider.
* One test for comparing timestamps at different before values, with the same data provider.
* One test for checking the ordering of timestamps, with a data provider that returns pairs of timestamps.

Note that the data providers can call each other; for example, given a `provideTimestampsWithoutSign` (same as your `timestampWithoutSignProvider`), you could have:

```php
public function provideTimestamps() { // any order
    foreach ( [ '+', '-' ] as $sign ) {
        foreach ( $this->provideTimestampsWithoutSign() as $timestamp ) {
            yield [ $sign . $timestamp ];
        }
    }
}

public function provideTimestampsAndPrecisions() {
    foreach ( $this->provideTimestamps() as $timestamp ) {
        foreach ( $this->providePrecisions() as $precision ) {
            yield [ $timestamp, $precision ];
        }
    }
}

public function provideTimestampPairs() {
    foreach ( [ '+', '-' ] as $sign ) {
        $oldTimestamp = null;
        foreach ( $this->provideTimestampsWithoutSign() as $timestamp ) {
            if ( $oldTimestamp !== null ) {
                yield $sign === '+' ? [ $oldTimestamp, $timestamp ] : [ $timestamp, $oldTimestamp ];
            }
            $oldTimestamp = $timestamp;
        }
    }
}
```

Same thing goes for the `getHigherTimestamp` tests (they could probably reuse the same data providers).